### ISSUE TITLE
version 2.0.3 broke google auth

### ISSUE THREAD

**Issue by veneliniliev** [spatie/laravel-google-cloud-storage#25]
When I use the latest version of the package (2.0.3) i have the following problem:

```
Google\Cloud\Core\Exception\BadRequestException
<?xml version='1.0' encoding='UTF-8'?><Error><Code>InvalidAuthentication</Code><Message>The provided authentication header is invalid.</Message><Details>Cannot use OAuth Authorization header with form POST.</Details></Error>
```

When reverting to 2.0.2... all work fine.

**Comment by freekmurze** [comment]
@hotrush can you take a look at this?

**Comment by hotrush** [comment]
@veneliniliev share please your config file (without secrets of course, just structure)

**Comment by hotrush** [comment]
Tested:

1. with path to config file
```
        'gcs' => [
            'driver' => 'gcs',
            'key_file_path' => '/app/gcs.json',
            'key_file' => [],
            'projectId' => null,
            'bucket' => 'BUCKET',
            'pathPrefix' => '',
            'apiEndpoint' => null,
            'visibility' => 'private',
            'metadata' => ['cacheControl'=> 'public,max-age=86400'],
        ],
```

2. with copied config file 
```
        'gcs' => [
            'driver' => 'gcs',
            'key_file_path' => null,
            'key_file' => [
                'type' => 'service_account',
                'project_id' => 'alien-device-123123',
                'private_key_id' => '12312',
                'private_key' => "-----BEGIN PRIVATE KEY-----...",
                'client_email' => '123.iam.gserviceaccount.com',
                'client_id' => '123',
                'auth_uri' => 'https://accounts.google.com/o/oauth2/auth',
                'token_uri' => 'https://oauth2.googleapis.com/token',
                'auth_provider_x509_cert_url' => 'https://www.googleapis.com/oauth2/v1/certs',
                'client_x509_cert_url' => 'https://www.googleapis.com/robot/v1/metadata/x...'
            ],
            'projectId' => null,
            'bucket' => 'BUCKET',
            'pathPrefix' => '',
            'apiEndpoint' => null,
            'visibility' => 'private',
            'metadata' => ['cacheControl'=> 'public,max-age=86400'],
        ],
```

3. with exported env var

`export GOOGLE_APPLICATION_CREDENTIALS=/app/gcs.json` plus

```
        'gcs' => [
            'driver' => 'gcs',
            'key_file_path' => null,
            'key_file' => [],
            'projectId' => null,
            'bucket' => 'BUCKET',
            'pathPrefix' => '',
            'apiEndpoint' => null,
            'visibility' => 'private',
            'metadata' => ['cacheControl'=> 'public,max-age=86400'],
        ],
```

Works perfect to me...

**Comment by veneliniliev** [comment]
it also works well for me locally.
the interesting thing is that when I built it on the docker image and put it in the kubernetes cluster then the problem happens.

**Comment by hotrush** [comment]
Might help if you tell us how you configure the package 😄 

Do you use GKE and creds from automatically mounted `GOOGLE_APPLICATION_CREDENTIALS`?

**Comment by veneliniliev** [comment]
i don't use `GOOGLE_APPLICATION_CREDENTIALS`

in filesystems
```
 'gcs' => [
            'driver' => 'gcs',
            'key_file_path' => resource_path(env('GOOGLE_CLOUD_KEY_FILE', null)), // optional: /path/to/service-account.json
            //'key_file' => [], // optional: Array of data that substitutes the .json file (see below)
            'project_id' => env('GOOGLE_CLOUD_PROJECT_ID', 'your-project-id'), // optional: is included in key file
            'bucket' => env('GOOGLE_CLOUD_STORAGE_BUCKET', 'your-bucket'),
            'path_prefix' => env('GOOGLE_CLOUD_STORAGE_PATH_PREFIX', ''), // optional: /default/path/to/apply/in/bucket
            'apiEndpoint' => env('GOOGLE_CLOUD_STORAGE_API_URI', null), // see: Public URLs below
            'visibility' => 'public', // optional: public|private
            'metadata' => ['cacheControl' => 'public,max-age=86400'], // optional: default metadata
        ],
```
in .env
```
GOOGLE_CLOUD_PROJECT_ID=xxx
GOOGLE_CLOUD_KEY_FILE="xxx/service-account.json"
GOOGLE_CLOUD_STORAGE_BUCKET=xxx
GOOGLE_CLOUD_STORAGE_API_URI="https://xxxx/"
```
env loaded correctly in pod
```
# env | grep 'GOOGLE'
GOOGLE_CLOUD_STORAGE_API_URI="https://xxxx/"
GOOGLE_CLOUD_STORAGE_BUCKET=xxxx
GOOGLE_CLOUD_PROJECT_ID=xxxx
GOOGLE_CLOUD_KEY_FILE="xxxx/service-account.json"
```

the configuration is the same as in 2.0.2 where it works without a problem.

**Comment by hotrush** [comment]
Changes i made fixes passing `apiEndpoint` into the client, so in your case i think it makes sense to remove `GOOGLE_CLOUD_STORAGE_API_URI="https://xxxx/"`. It doesn't make sense to override it usually, only when you use some side gcs-compatible storage.

If won't help - some questions:

- Why do you override default api url? 
- Is your `GOOGLE_CLOUD_PROJECT_ID` matching with project id from creds file?
- Key file path is relative? Makes sense to check if that correct

**Comment by shawnlindstrom** [comment]
@hotrush I can confirm removing the GOOGLE_CLOUD_STORAGE_API_URI from env resolved this issue for me.

**Comment by veneliniliev** [comment]
> @hotrush I can confirm removing the GOOGLE_CLOUD_STORAGE_API_URI from env resolved this issue for me.

resolve the issue but if you use a custom domain for cloud storage?

without this parameter, the domain is `https://storage.googleapis.com/...`
but I want to use my custom domain ex `https://cdn.custom-domain.com/...`

how to set my own domain?

**Comment by sergiofalcon** [comment]
Same problem of @veneliniliev here, and with a custom domain as bucket name, too. 

After several hours trying to fix it, it worked when i downgraded to 2.0.2, too.
